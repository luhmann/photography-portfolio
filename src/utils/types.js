import { arrayOf, shape, string, number } from 'prop-types';

export const locationType = shape({
  pathname: string.isRequired,
});

export const galleryInfoType = shape({
  title: string.isRequired,
  path: string.isRequired,
});

export const albumInfoType = shape({
  albumTitle: string.isRequired,
  galleries: arrayOf(galleryInfoType).isRequired,
});

export const imageType = shape({
  contentDigest: string.isRequired,
  fluid: shape({
    tracedSVG: string.isRequired,
    aspectRatio: number.isRequired,
    src: string.isRequired,
    srcSet: string.isRequired,
    srcWebp: string.isRequired,
    srcSetWebp: string.isRequired,
    sizes: string.isRequired,
  }).isRequired,
});

export const createImageTypeTestProps = (index, custom = {}) => ({
  contentDigest: `digest-${index}`,
  fluid: {
    tracedSVG: `data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' width='400' height='267' viewBox='0 0 400 267' version='1'%3e%3cpath d='tracedsvg-${index}' fill='lightgray' fill-rule='evenodd'/%3e%3c/svg%3e`,
    aspectRatio: 1.5,
    src: `/static/digest-${index}/size/mock-${index}.jpg`,
    srcSet: `/static/digest-${index}/size/mock-${index}-625.jpg 625w,\n/static/digest-${index}/size/mock-${index}-1250.jpg 1250w,\n/static/digest-${index}/size/mock-${index}-2250.jpg 2250w`,
    srcWebp: `/static/digest-${index}/size/mock-${index}.webp`,
    srcSetWebp: `/static/digest-${index}/size/mock-${index}-625.webp 625w,\n/static/digest-${index}/size/mock-${index}-1250.webp 1250w,\n/static/digest-${index}/size/mock-${index}-2250.webp 2250w`,
    sizes: '(max-width: 2250px) 100vw, 2250px',
  },
  ...custom,
});
