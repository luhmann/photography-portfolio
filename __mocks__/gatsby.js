const React = require('react');
const gatsby = jest.requireActual('gatsby');

const navigate = jest.fn();

module.exports = {
  ...gatsby,
  graphql: jest.fn(),
  navigate,
  Link: jest.fn().mockImplementation(({ to, ...rest }) =>
    React.createElement('a', {
      ...rest,
      href: to,
    })
  ),
  StaticQuery: jest.fn(),
};
