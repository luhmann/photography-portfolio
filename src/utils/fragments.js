export const GalleryImagesFragment = graphql`
  fragment GalleryImagesFragment on FileConnection {
    edges {
      node {
        childImageSharp {
          internal {
            contentDigest
          }
          sizes(maxWidth: 2500, quality: 75) {
            ...GatsbyImageSharpSizes_withWebp_tracedSVG
          }
        }
      }
    }
  }
`;

export const allGalleriesYamlFragment = graphql`
  fragment allGalleriesYamlFragment on RootQueryType {
    allGalleriesYaml {
      group(field: album) {
        fieldValue
        edges {
          node {
            title
            path
          }
        }
      }
    }
  }
`;

export const singleGalleryYamlFragment = graphql`
  fragment singleGalleryYamlFragment on RootQueryType {
    galleriesYaml(folderName: { eq: $folderName }) {
      title
    }
  }
`;
