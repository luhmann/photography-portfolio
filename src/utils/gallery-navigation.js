import { inc, dec, ifElse } from 'rambda';

const nextStepper = (index, total = 0) =>
  ifElse(() => inc(index) < total, () => inc(index), () => 0);

const prevStepper = (index, total = 0) =>
  ifElse(() => dec(index) >= 0, () => dec(index), () => dec(total));

export { nextStepper, prevStepper };
