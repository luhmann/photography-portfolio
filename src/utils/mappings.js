import { path, pathOr, pipe, map } from 'rambda';

const flattenAllGalleriesGraphQLResponse = response =>
  response.map(({ fieldValue: albumTitle, edges: galleries }) => ({
    albumTitle,
    galleries: galleries.map(path('node')),
  }));

export const mapAllGalleriesGraphQLResponse = pipe(
  pathOr([], 'allGalleriesYaml.group'),
  flattenAllGalleriesGraphQLResponse
);

export const mapGalleryImagesGraphQLResponse = pipe(
  pathOr([], 'data.allFile.edges'),
  map(
    pipe(
      pathOr([], 'node.childImageSharp'),
      img => ({
        fluid: img.fluid,
        contentDigest: path('internal.contentDigest', img),
      })
    )
  )
);

export const mapSingleGalleryYamlGraphQLResponse = path(
  'data.galleriesYaml.title'
);
