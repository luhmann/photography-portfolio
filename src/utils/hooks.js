import { useState, useEffect, useRef } from 'react';
import { pipe, path, find, findIndex, equals, head, defaultTo } from 'rambda';
import { navigate } from 'gatsby';

import { nextStepper, prevStepper } from 'utils/gallery-navigation';

export function useInterval(callback, delay) {
  const savedCallback = useRef();

  // remember the latest callback.
  useEffect(() => {
    savedCallback.current = callback;
  });

  // set up the interval.
  useEffect(
    () => {
      function tick() {
        savedCallback.current();
      }
      if (delay !== null) {
        let id = setInterval(tick, delay);
        return () => clearInterval(id);
      }
    },
    [delay]
  );
}

const getImageId = path('contentDigest');
const getCurrentIndexById = id => images =>
  findIndex(
    pipe(
      getImageId,
      equals(id)
    )
  )(images);

export function useGallery({ images = [], initialId }) {
  const total = images.length;
  const verifiedInitialId = pipe(
    find(
      pipe(
        getImageId,
        equals(initialId)
      )
    ),
    defaultTo(head(images)),
    getImageId
  )(images);
  const [id, setId] = useState(verifiedInitialId);

  useEffect(
    () => {
      setId(verifiedInitialId);
    },
    [verifiedInitialId]
  );

  if (total === 0) {
    navigate('/404');
    return {};
  }

  const getNextIndex = nextStepper(getCurrentIndexById(id)(images), total);
  const getPrevIndex = prevStepper(getCurrentIndexById(id)(images), total);
  const nextId = getImageId(images[getNextIndex()]);
  const prevId = getImageId(images[getPrevIndex()]);

  const next = () => setId(nextId);
  const prev = () => setId(prevId);

  return { next, prev, currentId: id, nextId, prevId };
}
