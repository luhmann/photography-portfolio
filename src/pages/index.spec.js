import React from 'react';
import { render, fireEvent, act } from 'react-testing-library';
import generateProps from 'react-generate-props';
import { StaticQuery } from 'gatsby';

import { createImageTypeTestProps } from 'utils/types';

import { IndexPage, IMAGE_DISPLAY_DURATION } from './index';

jest.useFakeTimers();

const createTestProps = custom => ({
  ...generateProps(IndexPage, { optional: true }),
  images: [
    createImageTypeTestProps(1),
    createImageTypeTestProps(2),
    createImageTypeTestProps(3),
    createImageTypeTestProps(4),
    createImageTypeTestProps(5),
    createImageTypeTestProps(6),
    createImageTypeTestProps(7),
  ],
  ...custom,
});

const testSelectors = Object.freeze({
  altImages: /Slideshow Image/,
  testIdEnterButton: 'home-enter-button',
  testIdLogo: 'home-logo',
});

beforeEach(() => {
  StaticQuery.mockImplementation(({ render }) => render({}));
});

test('should render as expected', () => {
  const props = createTestProps();
  const { getAllByAltText, getByTestId } = render(<IndexPage {...props} />);

  expect(getAllByAltText(testSelectors.altImages).length).toEqual(
    props.images.length
  );
  expect(getByTestId(testSelectors.testIdEnterButton)).toBeInTheDocument();
  expect(getByTestId(testSelectors.testIdLogo)).toBeInTheDocument();
});

test('should advance to the next image', () => {
  const props = createTestProps();
  const { asFragment } = render(<IndexPage {...props} />);

  const firstRender = asFragment();

  act(() => {
    jest.advanceTimersByTime(IMAGE_DISPLAY_DURATION + 1000);
  });

  expect(firstRender).toMatchDiffSnapshot(asFragment());
});
