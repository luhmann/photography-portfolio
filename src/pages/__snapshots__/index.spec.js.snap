// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should advance to the next image 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -24,11 +24,11 @@
      />
      <div
        class=\\"sc-bwzfXH iVSscL\\"
      >
        <div
-         class=\\"sc-gzVnrw eGhSyp\\"
+         class=\\"sc-gzVnrw bfIvDc\\"
        >
          <div
            class=\\" gatsby-image-wrapper\\"
            style=\\"position: relative; overflow: hidden; height: 100%;\\"
          >
@@ -68,11 +68,11 @@
  /static/digest-1/size/mock-1-2250.jpg 2250w\\" sizes=\\"(max-width: 2250px) 100vw, 2250px\\" /&gt;&lt;img src=\\"/static/digest-1/size/mock-1.jpg\\" alt=\\"Slideshow Image-1\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/&gt;&lt;/picture&gt;
            </noscript>
          </div>
        </div>
        <div
-         class=\\"sc-gzVnrw bfIvDc\\"
+         class=\\"sc-gzVnrw eGhSyp\\"
        >
          <div
            class=\\" gatsby-image-wrapper\\"
            style=\\"position: relative; overflow: hidden; height: 100%;\\"
          >"
`;
